feature: Allow str arguments in Trainer.profiler

ddrevicky opened the pull request:
## What does this PR do?

<!--
Please include a summary of the change and which issue is fixed.
 Please also include relevant motivation and context.
 List any dependencies that are required for this change.
-->
Fixes #3330 

- Allows `str` parameters for `profiler` arg in `Trainer`
- Deprecates `bool` parameters for `profiler` arg in `Trainer`
- Removes `bool` parameters for `profiler` from examples in docs
- Adds tests checking correct profilers are returned for `str` arguments

```
# equivalent options
# CLI: --profiler true (deprecated) or --profiler simple
trainer = Trainer(profiler=True)        # deprecated
trainer = Trainer(profiler=SimpleProfiler())
trainer = Trainer(profiler="simple")

# equivalent options
# CLI: --profiler advanced
trainer = Trainer(profiler=AdvancedProfiler())
trainer = Trainer(profiler="advanced")
```

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

<!-- For CHANGELOG separate each item in unreleased section by a blank line to reduce collisions -->

pep8speaks commented:
Hello @ddrevicky! Thanks for opening this PR. 

* In the file [`pytorch_lightning/profiler/__init__.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/3023402b34245b2c6426a76f5fd016e4e1774e38/pytorch_lightning/profiler/__init__.py):

> [Line 63:1](https://github.com/PyTorchLightning/pytorch-lightning/blob/3023402b34245b2c6426a76f5fd016e4e1774e38/pytorch_lightning/profiler/__init__.py#L63): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace









* In the file [`tests/trainer/test_trainer.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/3023402b34245b2c6426a76f5fd016e4e1774e38/tests/trainer/test_trainer.py):

> [Line 1145:1](https://github.com/PyTorchLightning/pytorch-lightning/blob/3023402b34245b2c6426a76f5fd016e4e1774e38/tests/trainer/test_trainer.py#L1145): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1149:43](https://github.com/PyTorchLightning/pytorch-lightning/blob/3023402b34245b2c6426a76f5fd016e4e1774e38/tests/trainer/test_trainer.py#L1149): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent

Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

justusschock commented on pytorch_lightning/trainer/connectors/profiler_connector.py:
```diff
@@ -11,16 +11,24 @@
+    def on_trainer_init(self, profiler: Union[BaseProfiler, bool, str]):
         # configure profiler
-        if profiler is True:
+        if profiler is True or profiler == "simple":
```
we should not have two options to do this. IMO we should then always go with a string. Thoughts @Borda ?

ddrevicky replied:
So backwards incompatibility would not be a problem?

justusschock replied:
that's a good point. Probably we need to keep it then but deprecate it?

ddrevicky replied:
Just checking:
All explicitly boolean arg values (True/False) would trigger a deprecation warning right? Also, currently when running from the CLI, one can do `my_file.py --profiler` which would be also deprecated in this case and `my_file.py --profiler simple` would be the new way of doing things.

Borda replied:
I feel like a string is quite sensitive to typos, so what about Enumn?

justusschock replied:
this is not intuitive for users with argparse

Borda replied:
```
from enum import Enum

class Prof(Enum):
    ABC = 'abc'
    BCD = 'bcd'

print (Prof.ABC)
print (Prof('abc'))
```

justusschock replied:
but then from the argparse it would be a string again? So still typos? I don't really think an enum is necessary here

ddrevicky replied:
If using an enum it would be best to also change `--auto_scale_batch_size` which also [uses strings](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/tuner/batch_size_scaling.py#L100) and [--auto_lr_find](https://github.com/PyTorchLightning/pytorch-lightning/blob/1d7c615d82588c6b01b8926a2afd346c8cc3b8ad/pytorch_lightning/tuner/lr_finder.py#L243) as well.

ddrevicky replied:
@Borda what do you think?

Borda replied:
> but then from the argparse it would be a string again? So still typos? I don't really think an enum is necessary here

yes but it will we caught as soon as you make that casting not later while running

> If using an enum it would be best to also change `--auto_scale_batch_size` which also [uses strings](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/tuner/batch_size_scaling.py#L100) and [--auto_lr_find](https://github.com/PyTorchLightning/pytorch-lightning/blob/1d7c615d82588c6b01b8926a2afd346c8cc3b8ad/pytorch_lightning/tuner/lr_finder.py#L243) as well.

yes, a string is ine

carmocca commented on tests/trainer/test_trainer.py:
```diff
@@ -1127,3 +1128,23 @@ def setup(self, stage):
     assert trainer.get_model().stage == 'test'
+
+
+@pytest.mark.parametrize("trainer_kwargs,expected", [
+    pytest.param(dict(), PassThroughProfiler),
+    pytest.param(dict(profiler=False), PassThroughProfiler),
+    pytest.param(dict(profiler=True), SimpleProfiler),
+    pytest.param(dict(profiler=SimpleProfiler()), SimpleProfiler),
+    pytest.param(dict(profiler=AdvancedProfiler()), AdvancedProfiler),
+    pytest.param(dict(profiler="simple"), SimpleProfiler),
+    pytest.param(dict(profiler="advanced"), AdvancedProfiler),
+])
+def test_trainer_profiler_correct_args(trainer_kwargs, expected):
+    trainer = Trainer(**trainer_kwargs)
```
I think this is cleaner :smile: 

```suggestion
@pytest.mark.parametrize(['input', 'expected'], [
    (None, PassThroughProfiler),
    (False, PassThroughProfiler),
    (True, SimpleProfiler),
    (SimpleProfiler(), SimpleProfiler),
    (AdvancedProfiler(), AdvancedProfiler),
    ("simple", SimpleProfiler),
    ("advanced", AdvancedProfiler),
])
def test_trainer_profiler_correct_args(input, expected):
    kwargs = {'profiler': input} if input else {}
    trainer = Trainer(**kwargs)
```

ddrevicky replied:
yep, much cleaner thanks :)

carmocca commented on CHANGELOG.md:
```diff
@@ -35,6 +37,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Changed IoU `remove_bg` bool to `ignore_index` optional int ([#3098](https://github.com/PyTorchLightning/pytorch-lightning/pull/3098))
 
+
```
Extra line?

ddrevicky commented:
I've added deprecation of bool argument for the profiler like @justusschock requested. I'm not sure I've done all the steps for deprecation correctly, especially:

- version in which removal occurs
- deprecation tests in `test_deprecated.py`
- I removed examples of passing bool to profiler from docs, is that right?

Let me know if there's anything that needs correction. The failing checks seem to be unrelated to the PR.

ydcjeff commented on pytorch_lightning/trainer/connectors/profiler_connector.py:
```diff
@@ -11,16 +11,30 @@
             profiler = SimpleProfiler()
+        elif profiler == "advanced":
+            profiler = AdvancedProfiler()
+        elif isinstance(profiler, str):
```
Shall we also not allow other types?

rohitgr7 replied:
yeah I think we can have a condition above
```python
if not isinstance(profiler, (bool, str, BaseProfiler)):
    raise MisconfigurationException(...)
```

ydcjeff replied:
or rather replace
```py
elif isinstance(profiler, str):
```
with
```py
if not isinstance(profiler, (bool, 'simple', 'advanced', BaseProfiler)):
    raise MisconfigurationException(...)
```
if we only allow `simple` and `advanced`

ddrevicky replied:
@ydcjeff Are you sure the `isinstance` construct is valid? Python 3.8 throws me a `TypeError`:
`>>> isinstance("abcd", (bool, "abcd"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: isinstance() arg 2 must be a type or tuple of types
`

Borda replied:
@ydcjeff `'simple', 'advanced'` this is not data type...

ydcjeff replied:
Ahh yeah, sorry for that

ydcjeff replied:
Can we have a test for passing other types than string?

ddrevicky replied:
Added a test for incorrect datatypes (int, float, dict, list, tensor, Trainer) for which a `MisconfigurationException` is raised. A previous test already checks that the Profiler is created as expected for correct argument types.

ydcjeff commented:
You may want to add warning of `bool` [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/43ec93fb6a7d18a30a3e380aaa9fd567de992eb0/pytorch_lightning/trainer/trainer.py#L204)

rohitgr7 commented on pytorch_lightning/trainer/connectors/profiler_connector.py:
```diff
@@ -11,16 +11,30 @@
+        elif profiler == "advanced":
+            profiler = AdvancedProfiler()
+        elif isinstance(profiler, str):
+            raise ValueError('when passing string value for the `profiler` parameter of'
+                             ' `Trainer`, it can only be `simple` or `advanced`')
```
```suggestion
            raise ValueError("When passing string value for the `profiler` parameter of"
                             " `Trainer`, it can only be 'simple' or 'advanced'")
```

rohitgr7 commented on tests/test_deprecated.py:
```diff
@@ -27,6 +28,16 @@ def test_tbd_remove_in_v0_11_0_trainer_gpu():
         gpu_usage = GpuUsageLogger()
 
 
+@pytest.mark.parametrize(['input_arg', 'expected'], [
+    (True, SimpleProfiler),
+    (False, PassThroughProfiler),
+])
+def test_trainer_profiler_remove_in_v0_11_0_trainer(input_arg, expected):
+    with pytest.deprecated_call(match='will be removed in v0.11.0'):
+        trainer = Trainer(profiler=input_arg)
+        assert isinstance(trainer.profiler, expected)
```
```suggestion
@pytest.mark.parametrize(['profiler', 'expected'], [
    (True, SimpleProfiler),
    (False, PassThroughProfiler),
])
def test_trainer_profiler_remove_in_v0_11_0_trainer(profiler, expected):
    with pytest.deprecated_call(match='will be removed in v0.11.0'):
        trainer = Trainer(profiler=profiler)
        assert isinstance(trainer.profiler, expected)
```

rohitgr7 commented on tests/trainer/test_trainer.py:
```diff
@@ -1179,3 +1180,22 @@ def test_row_log_interval(log_metrics_mock, tmpdir, train_batches, max_steps, lo
+
+
+def test_trainer_profiler_incorrect_str_arg():
+    with pytest.raises(ValueError, match=r'when passing string value for the `profiler` parameter of '
+                       '`Trainer`, it can only be `simple` or `advanced`'):
```
```suggestion
    with pytest.raises(ValueError, match=r"When passing string value for the `profiler` parameter of"
                       "` Trainer`, it can only be 'simple' or 'advanced'"):
```

rohitgr7 commented on tests/trainer/test_trainer.py:
```diff
@@ -1179,3 +1180,22 @@ def test_row_log_interval(log_metrics_mock, tmpdir, train_batches, max_steps, lo
     log_metrics_mock.assert_has_calls(expected_calls)
+
+    
+@pytest.mark.parametrize(['input_arg', 'expected'], [
+    (None, PassThroughProfiler),
+    (SimpleProfiler(), SimpleProfiler),
+    (AdvancedProfiler(), AdvancedProfiler),
+    ('simple', SimpleProfiler),
+    ('advanced', AdvancedProfiler),
+])
+def test_trainer_profiler_correct_args(input_arg, expected):
+    kwargs = {'profiler': input_arg} if input_arg is not None else {}
+    trainer = Trainer(**kwargs)
```
```suggestion
@pytest.mark.parametrize(['profiler', 'expected'], [
    (None, PassThroughProfiler),
    (SimpleProfiler(), SimpleProfiler),
    (AdvancedProfiler(), AdvancedProfiler),
    ('simple', SimpleProfiler),
    ('advanced', AdvancedProfiler),
])
def test_trainer_profiler_correct_args(profiler, expected):
    trainer = Trainer(profiler=profiler)
```

ddrevicky replied:
Thanks for the suggestion, I kept the `kwargs` since this way it also checks if `None` is used as default if nothing is supplied as the `profiler` arg.

rohitgr7 commented on pytorch_lightning/trainer/connectors/profiler_connector.py:
```diff
@@ -11,16 +11,30 @@
+                           DeprecationWarning)
         # configure profiler
-        if profiler is True:
+        if profiler is True or profiler == "simple":
             profiler = SimpleProfiler()
+        elif profiler == "advanced":
```
should we allow uncased values like `Simple` and `Advanced` too? I mean it's just one linear.

justusschock replied:
yeah, we should simply cast profiler to lower if it's a string

ddrevicky replied:
updated

Borda replied:
cases are very easy to make typo :D

rohitgr7 reviewed (approved):
Just some suggestions. Btw LGTM. Nice work!

ddrevicky commented on tests/trainer/test_trainer.py:
```diff
@@ -1179,3 +1180,22 @@ def test_row_log_interval(log_metrics_mock, tmpdir, train_batches, max_steps, lo
+    ('advanced', AdvancedProfiler),
+])
+def test_trainer_profiler_correct_args(profiler, expected):
+    kwargs = {'profiler': profiler} if profiler is not None else {}
```
Thanks for the suggestions @rohitgr7. I kept the `kwargs` here since that also checks that the `profiler` argument is correctly defaulted to `None` in case nothing is passed. Also, your suggestion for this is no longer here I think it's because I force pushed an ammend (sorry :)).

ydcjeff commented on pytorch_lightning/trainer/trainer.py:
```diff
@@ -201,7 +201,8 @@ def __init__(
                 Ignored when a custom callback is passed to :paramref:`~Trainer.callbacks`.
 
-            profiler:  To profile individual steps during training and assist in identifying bottlenecks.
+            profiler:  To profile individual steps during training and assist in identifying bottlenecks. Passing bool
+                value is deprecated and will be removed in 0.11.0.
```
```suggestion
            profiler:  To profile individual steps during training and assist in identifying bottlenecks.
            
                .. warning::
                    Passing bool value is deprecated and will be removed in 0.11.0.
```

Borda replied:
```suggestion
                value is deprecated in v0.9.1 and will be removed in 0.11.0.
```

carmocca commented on pytorch_lightning/trainer/connectors/profiler_connector.py:
```diff
@@ -11,16 +11,40 @@
+    def on_trainer_init(self, profiler: Union[BaseProfiler, bool, str]):
+
+        if profiler and not isinstance(profiler, (bool, str, BaseProfiler)):
+            # TODO: Update exception on deprecation of bool
```
You already deprecated bool, I think you meant:

```suggestion
            # TODO: Update exception on removal of bool
```

carmocca commented on pytorch_lightning/trainer/trainer.py:
```diff
@@ -201,7 +201,8 @@ def __init__(
 
-            profiler:  To profile individual steps during training and assist in identifying bottlenecks.
+            profiler:  To profile individual steps during training and assist in identifying bottlenecks. Passing bool
+                value is deprecated and will be removed in 0.11.0.
```
AFAIK, there will be 0.10.0 and then 1.0 with all the deprecated stuff removed.

Can you confirm @edenafek ?

ddrevicky replied:
so "will be removed in 1.0" correct? I was just guessing based on [existing tests](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/test_deprecated.py#L19) in `test_deprecated.py` which mention `v0.11.0`

rohitgr7 replied:
asked on slack, it's not a problem. All deprecations after 0.10.0 will be removed before 1.0.

Borda replied:
let's write there 0.11 we already have some other noted as 0.11 so keep it consistent :]

ddrevicky replied:
Seems ready then. 

Failing checks:
- Speed comparison with PyTorch is failing `continuous-integration/drone/pr` because of a few milliseconds `AssertionError: lightning [0.55572604 0.37785082] was slower than PT (threshold 0.55)`
-  `CodeFactor` doesn't like TODOs

carmocca commented:
LGTM. Great job! :muscle:

Borda commented on pytorch_lightning/trainer/__init__.py:
```diff
@@ -835,14 +835,11 @@ def on_train_end(self, trainer, pl_module):
-    trainer = Trainer(profiler=SimpleProfiler())
-
-    # advanced profiler for function-level stats
-    trainer = Trainer(profiler=AdvancedProfiler())
```
what happened with these, can we still pass Profiler object?

Borda commented on pytorch_lightning/trainer/trainer.py:
```diff
@@ -117,7 +117,7 @@ def __init__(
         truncated_bptt_steps: Optional[int] = None,
         resume_from_checkpoint: Optional[str] = None,
-        profiler: Optional[Union[BaseProfiler, bool]] = None,
+        profiler: Optional[Union[BaseProfiler, bool, str]] = None,
```
```suggestion
        profiler: Optional[Union[BaseProfiler, str]] = None,
```

Borda commented on pytorch_lightning/trainer/__init__.py:
```diff
@@ -835,14 +835,11 @@ def on_train_end(self, trainer, pl_module):
 
-    # to profile standard training events
-    trainer = Trainer(profiler=True)
+    # to profile standard training events, equivalent to profiler=SimpleProfiler()
```
```suggestion
    # to profile standard training events, equivalent to `profiler=SimpleProfiler()`
```

Borda commented on pytorch_lightning/trainer/__init__.py:
```diff
@@ -835,14 +835,11 @@ def on_train_end(self, trainer, pl_module):
-
-    # advanced profiler for function-level stats
-    trainer = Trainer(profiler=AdvancedProfiler())
+    # advanced profiler for function-level stats, equivalent to profiler=AdvancedProfiler()
```
```suggestion
    # advanced profiler for function-level stats, equivalent to `profiler=AdvancedProfiler()`
```

awaelchli commented on pytorch_lightning/utilities/argparse_utils.py:
```diff
@@ -91,6 +91,7 @@ def get_init_arguments_and_types(cls) -> List[Tuple[str, Tuple, Any]]:
          ('profiler',
           (<class 'pytorch_lightning.profiler.profilers.BaseProfiler'>,
            <class 'bool'>,
+           <class 'str'>,
```
```suggestion
           <class 'str'>,
```

I think this is why tests fail. Two options:
1. include bool in the type annotation of Trainer
2. remove the bool here. 

Either way, the signature of Trainer must match what is listed here.

ddrevicky replied:
I went with option no. 1, see [commit](https://github.com/PyTorchLightning/pytorch-lightning/pull/3656/commits/d7b2e7449d883a425fe0f7ee8da7199dc45eae84). 

`bool` needs to stay in the `profiler` type in `trainer.py` because , `Trainer.add_argparse_args(parent_parser=parser)` deduces from the types of the `Trainer`'s `profiler` parameter what type the argument added to `ArgParser` should have. If `bool` is not in the signature of the `profiler` parameter, the type of the argument added to `ArgParser` is `str`.

Then if CLI args like "--profiler True", the value of the parsed arg would be `"True"` instead of `True` and `Trainer` would raise an exception complaining that only `"simple"` and `"advanced"` are acceptable `str` values.

ddrevicky replied:
@awaelchli could you please take a look at this PR? If you are satisfied with the changes, I think it can be merged :)

awaelchli replied:
yes, this seems to be resolved now. 
the PR does not have a milestone, so I don't know if this will go into v1.0. 
need to ask @williamFalcon @edenafek

ddrevicky replied:
Okay, thanks I haven't realized that.

carmocca commented:
Should this PR not deprecate and just break compat considering 1.0 already does?

awaelchli commented on pytorch_lightning/trainer/trainer.py:
```diff
@@ -212,7 +212,8 @@ def __init__(
                 Ignored when a custom callback is passed to :paramref:`~Trainer.callbacks`.
 
-            profiler:  To profile individual steps during training and assist in identifying bottlenecks.
+            profiler:  To profile individual steps during training and assist in identifying bottlenecks. Passing bool
```
```suggestion
            profiler: To profile individual steps during training and assist in identifying bottlenecks. Passing bool
```

awaelchli commented on pytorch_lightning/trainer/trainer.py:
```diff
@@ -212,7 +212,8 @@ def __init__(
 
-            profiler:  To profile individual steps during training and assist in identifying bottlenecks.
+            profiler:  To profile individual steps during training and assist in identifying bottlenecks. Passing bool
+                value is deprecated in v0.9.1 and will be removed in 0.11.0.
```
```suggestion
                value is deprecated in v1.1 and will be removed in v1.3.
```

awaelchli reviewed (commented):
can you check again all mentions of versions in deprecation messages and TODOs. 
next version will be 1.1, so deprecation should probably start from there and then dropping the support in v1.3? cc @Borda

rohitgr7 commented:
I think it should deprecate in 1.0 and remove in 1.2. That's what I did with filepath in ModelCheckpoint.

ddrevicky commented:
@Borda could you please confirm which of the above deprecation strategies is correct? Also is that a rule of thumb for smaller changes like this, that when deprecating in version 1.x, we drop support in 1.(x+2), e.g. 1.1 and 1.3? Thanks.

Borda commented:
> @Borda could you please confirm which of the above deprecation strategies is correct? Also is that a rule of thumb for smaller changes like this, that when deprecating in version 1.x, we drop support in 1.(x+2), e.g. 1.1 and 1.3? Thanks.

@rohitgr7 is this is set for 1.1 release and in that one, it will be released @awaelchli is correct

williamFalcon commented:
anything deprecated needs 2 minor releases. 

ie: if we deprecate something today, it should be removed about 4-5 months from now in 1.3.0

ddrevicky commented:
@awaelchli Okay I updated the deprecation messages and TODOs as suggested by @Borda. It now says that bool is deprecated in 1.1 and will be removed in 1.3. I am still somewhat confused by this, because since if we merge this now, bool as profiler param is deprecated and it's not version 1.1 yet. So shouldn't it say bool is deprecated in 1.0?
